provider/provider.go

```diff
@@ -0,0 +1,139 @@
+package provider
+
+import (
+	"context"
+	"errors"
+	"net"
+	"os"
+	"sync"
+
+	"github.com/codedellemc/gocsi"
+	"github.com/codedellemc/gocsi/csi"
+	"github.com/codedellemc/goioc"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+
+	"github.com/codedellemc/csi-blockdevices/services"
+)
+
+const (
+	nodeEnvVar = "BDPLUGIN_NODEONLY"
+	ctlrEnvVar = "BDPLUGIN_CONTROLLERONLY"
+)
+
+var (
+	errServerStarted = errors.New(
+		services.SpName + ": the server has been started")
+	errServerStopped = errors.New(
+		services.SpName + ": the server has been stopped")
+)
+
+func init() {
+	goioc.Register(services.SpName, newProvider)
+}
+
+type provider struct {
+	sync.Mutex
+	server *grpc.Server
+	closed bool
+	plugin *services.StoragePlugin
+}
+
+func newProvider() interface{} {
+	return &provider{}
+}
+
+// Serve accepts incoming connections on the listener lis, creating
+// a new ServerTransport and service goroutine for each. The service
+// goroutine read gRPC requests and then call the registered handlers
+// to reply to them. Serve returns when lis.Accept fails with fatal
+// errors.  lis will be closed when this method returns.
+// Serve always returns non-nil error.
+func (p *provider) Serve(ctx context.Context, li net.Listener) error {
+	log.WithField("name", services.SpName).Info(".Serve")
+	if err := func() error {
+		p.Lock()
+		defer p.Unlock()
+		if p.closed {
+			return errServerStopped
+		}
+		if p.server != nil {
+			return errServerStarted
+		}
+		p.server = grpc.NewServer(grpc.UnaryInterceptor(gocsi.ChainUnaryServer(
```

Comment:
Hi @codenrhoden,

Since you're following this same pattern in `main.go` and you didn't centralize the interceptors into the Provider, I'd recommend removing them from the Provider. The reason being is that if the provider is imported into REX-Ray then I'd expect REX-Ray to handle the logging and validation. Does that make sense?

Reply:
It does! I'm glad you caught that, as I meant to specifically ask about it -- "Do I need the interceptors here, or will REX-Ray do it?". :) Now I know.